src/app/archive_hardfork_toolbox/sql.ml

```diff
@@ -0,0 +1,196 @@
+open Core
+open Caqti_request.Infix
+
+module type CONNECTION = Mina_caqti.CONNECTION
+
+let latest_state_hash_query =
+  (Caqti_type.unit ->! Caqti_type.string)
+    {sql|
+          SELECT state_hash from blocks order by height desc limit 1;
+        |sql}
+
+let latest_state_hash (module Conn : CONNECTION) =
+  Conn.find latest_state_hash_query ()
+
+let is_in_the_best_chain_query =
+  (Caqti_type.(t4 string string int int64) ->! Caqti_type.bool)
+    {sql|
+    WITH RECURSIVE chain AS (
+      SELECT
+        b.id,
+        NULLIF(b.parent_id, 0) AS parent_id,
+        b.state_hash,
+        b.height,
+        b.global_slot_since_genesis,
+        ARRAY[b.id] AS path
+      FROM blocks b
+      WHERE b.state_hash = ?
+
+      UNION ALL
+
+      SELECT
+        p.id,
+        NULLIF(p.parent_id, 0) AS parent_id,
+        p.state_hash,
+        p.height,
+        p.global_slot_since_genesis,
+        c.path || p.id
+      FROM blocks p
+      JOIN chain c ON p.id = c.parent_id
+      WHERE NOT p.id = ANY (c.path)
+    )
+    SELECT EXISTS (
+      SELECT 1 FROM chain
+      WHERE state_hash = ?
+        AND height = ?
+        AND global_slot_since_genesis = ?
+    ) AS is_in_chain;
+    |sql}
+
+let is_in_the_best_chain (module Conn : CONNECTION) ~tip_hash ~check_hash
+    ~check_height ~check_slot () =
+  Conn.find is_in_the_best_chain_query
+    (tip_hash, check_hash, check_height, check_slot)
+
+let no_of_confirmations_query =
+  (Caqti_type.(t2 string int) ->! Caqti_type.int)
+    {sql|
+    WITH RECURSIVE chain AS
```

Comment:
NOTE: as a follow up. This recursive chain query showed up in multiple queries here. It's better to improve this by execute this query only once.